### DISCUSSION TITLE
question: should my Thunderbird Add-on support automatic css transformation?

### DISCUSSION THREAD

**Discussion by RealRaven2000** [email-markup-consortium/email-markup-consortium#97] [Q&A]
Hi, I have just  watched Kevin Powell's [interview with Mark Powell](https://youtu.be/MHnTpN0g7ko) and found the github here on his recommendation.
I am the author of the[ Thunderbird Add-on SmartTemplates](https://addons.thunderbird.net/thunderbird/addon/smarttemplate4/), which supports building rich HTML emails (that works well on Thunderbird). It has commands for injecting HTML fragments, style sheets and images to simplify building rich HTML email. While going down the rabbit hole of compatibility, one of the issues is support for `<style>` tags, which seem to be ignored by many clients. 

My question: would it make sense / be useful to implement a feature that iterates the DOM elements of the email and automatically injects all (resulting) rules from the style section into the individual elements, as something the user could choose to do before they send off the email?

**Comment by husseinalhammad** [comment]
Hi,

This is perhaps not the best repo to discuss this. Nonetheless, I realise not many alternatives exist so I'm keeping it.

> My question: would it make sense / be useful to implement a feature that iterates the DOM elements of the email and automatically injects all (resulting) rules from the style section into the individual elements, as something the user could choose to do before they send off the email?

Yes. My suggestion would be to use a CSS inliner like Juice (https://github.com/Automattic/juice).

You can still use embedded CSS in `<style>` for further enhancements (e.g. `:hover` states, refine styles at certain breakpoints), but with the awareness that these may not work everywhere. See: https://www.caniemail.com/features/html-style/

**Reply by RealRaven2000** [reply]
thanks for your answer. Juice look good I hope it hasn't got too many external dependencies, as I am trying to keep my Add-ons as lean as possible to avoid lengthy review times - the Add-on is still in "experimental" status as I find it hard to fully convert it into an API-based Add-on because of the big amounts of restrictions. (It used to be a legacy Add-on before Thunderbird 68)

I also have to look into possible problems with my license, currently I am considering to switch to a less restrictive, tailor made licensing model which should also help including other people's work better.

For the moment [I have merged](https://github.com/RealRaven2000/SmartTemplates/issues/79) `<head>` section items from the template and extracting `<body>` attributes from the template to insert into the `body` of the email (as the DOM of the composer document silently drops any `<body>` and `<head>` tags, while keeping the `innerHTML` contents in place. But inlining styles would seriously harden my user's layouts so this may be a better long term solution.
